docs: introduce build system docs (#8964)
backstage/backstage: docs/local-dev/cli-build-system.md

[loaders](#loaders) and [transpilation](#transpilation) sections.

The output of the bundling process is split into two categories of files with
separate caching strategies. The first is a set of generic assets with plain
names in the root of the `dist/` folder. You will want to serve these with
short-lived caching or no caching at all. The second is a set of hashed static
assets in the `dist/static/` folder, which you can configure to be cached for a
much longer time.

The configuration of static assets is optimized for frequent changes and serving
over HTTP 2.0. The assets are aggressively split into small chunks, which means
the browser has to make a lot of small requests to load them. The upside is that
changes to individual plugins and packages will invalidate a smaller number of
files, thereby allowing for rapid development without much impact on the page
load performance.

### Backend Development

The backend development bundling is also based on Webpack, but rather than
starting up a web server, the backend is started up using the
[`RunScriptWebpackPlugin`](https://www.npmjs.com/package/run-script-webpack-plugin).

> The reason for using Webpack for development of the backend is both that it is

Reviewer: ```suggestion
The reason for using Webpack for development of the backend is both that it is a
```

convenient way to handle transpilation of a large set of packages, as well us

> allowing us to use hot module replacement and maintain state while reloading

Reviewer: ```suggestion
allowing us to use hot module replacement and maintaining state while reloading
```

individual backend modules. This is particularly useful when running the backend
with in-memory SQLite as the database choice.

Except for executing in Node.js rather than a web server, the backend
development bundling configuration is quite similar to the frontend one. It
shares most of the Webpack configuration, including the transpilation setup.
Some differences are that it does not inject any environment variables or node
module fallbacks, and it uses
[`webpack-node-externals`](https://www.npmjs.com/package/webpack-node-externals)
to avoid bundling in dependency modules.

If you want to inspect the running Node.js process, the `--inspect` and
`--inspect-brk` flags can be used, as they will be passed through as options to
`node` execution.

### Backend Production

The backend production bundling uses a completely different setup than the other
bundling options. Rather than using Webpack, the backend production bundling

> instead collects the backend packages and all of its local dependencies into a

Reviewer: ```suggestion
instead collects the backend packages and all of their local dependencies into a
```

deployment archive. The archive is written to `dist/bundle.tar.gz`, and contains
the packaged version of each of these packages. The layout of the packages in
the archive is the same as the directory layout in the monorepo, and the bundle
also contains the root `package.json` and `yarn.lock` files.

Note that before creating a production bundle you must first build all of the
backend packages. This can be done automatically when executing the
`backend:bundle` command by passing the `--build-dependencies` flag. It is an

> optional flag since it is quite common that the packages are already build

Reviewer: ```suggestion
optional flag since it is quite common that the packages are already built
```

output of `backstage-cli backend:bundle` into an image:

```Dockerfile
FROM node:14-buster-slim
WORKDIR /app

COPY yarn.lock package.json packages/backend/dist/skeleton.tar.gz ./
RUN tar xzf skeleton.tar.gz && rm skeleton.tar.gz

RUN yarn install --production --frozen-lockfile --network-timeout 300000 && rm -rf "$(yarn cache dir)"

COPY packages/backend/dist/bundle.tar.gz app-config.yaml ./
RUN tar xzf bundle.tar.gz && rm bundle.tar.gz

CMD ["node", "packages/backend"]
```

## Transpilation

The transpilers used by the Backstage CLI have been chosen according to the same
design considerations that were mentioned above. A few specific requirements are
of course support for TypeScript and JSX, but also React hot reloads or refresh,
and hoisting of Jest mocks. The Backstage CLI also only targets up to date and
modern browsers, so we actually want to keep the transpilation process as
lightweight as possible, and leave most syntax intact.

Apart from these requirements, the deciding factor for which transpiler to use
is their speed. The build process keeps the integration with the transpilers
lightweight, without additional plugins or such. This enables us to switch out

> transpilers as new options and optimizations become available, and keep use the

Reviewer: ```suggestion
transpilers as new options and optimizations become available, and keep on using the
```

| `.gif`      | URL Path        | Image                                                                         |
| `.jpeg`     | URL Path        | Image                                                                         |
| `.jpg`      | URL Path        | Image                                                                         |
| `.png`      | URL Path        | Image                                                                         |
| `.svg`      | URL Path        | Image                                                                         |
| `.icon.svg` | React Component | SVG converted into a [MUI SvgIcon](https://mui.com/components/icons/#svgicon) |

## Jest Configuration

The Backstage CLI bundles its own Jest configuration file, which is used
automatically when running `backstage-cli test`. It's available at
`@backstage/cli/config/jest.js` and can be inspected
[here](https://github.com/backstage/backstage/blob/master/packages/cli/config/jest.js).
Usage of this configuration can be overridden either by passing a
`--config <path>` flag to `backstage-cli test`, or placing a `jest.config.js` or
`jest.config.ts` file in your package.

> The build-in configuration brings a couple of benefits and features. The most

Reviewer: ```suggestion
The built-in configuration brings a couple of benefits and features. The most
```

important one being a baseline transformer and module configuration that enables
support for the listed [loaders](#loaders) within tests. It will also

> automatically detect and uses `src/setupTests.ts` if it exists, and provide a

Reviewer: ```suggestion
automatically detect and use `src/setupTests.ts` if it exists, and provides a
```

coverage configuration that works well with our selected transpilers.

The configuration also takes a project-wide approach, with the expectation most
if not all packages within a monorepo will use the same base configuration. This
allows for optimizations such as sharing the Jest transform cache across all
packages in a monorepo, avoiding unnecessary transpilation. It also makes it
possible to load in all Jest configurations at once, and with that run
`yarn test <pattern>` from the root of a monorepo without having to set the
working directory to the package that the test is in.

Where small customizations are needed, such as setting coverage thresholds or
support for specific transforms, it is possible to override the Jest

> configuration through the `"jest"` in `package.json`. These overrides will be

Reviewer: ```suggestion
configuration through the `"jest"` field in `package.json`. These overrides will be
```

loaded in from all `package.json` files in the directory ancestry, meaning that
you can place common configuration in the `package.json` at the root of a
monorepo. If multiple overrides are found, they will be merged together with
configuration further down in the directory tree taking precedence.

The overrides in a single `package.json` may for example look like this:

```json
  "jest": {
    "coverageThreshold": {
      "global": {
        "functions": 100,
        "lines": 100,
        "statements": 100
      }
    }
  },
```

## Publishing

Package publishing is an optional part of the Backstage build system and not

> something you well need to worry unless you are publishing packages to a

Reviewer: ```suggestion
something you will need to worry about unless you are publishing packages to a
```

registry. In order to publish a package, you first need to build it, which will
populate the `dist` folder. Because the Backstage build system is optimized for
local development along with our particular TypeScript and bundling setup, it is

> not possible to publish package immediately at this point. This is because the

Reviewer: ```suggestion
not possible to publish the package immediately at this point. This is because the
```

entry points of the package will still be pointing to `src/index.ts`, but we
want them to point to `dist/` in the published package.

In order to work around this, the Backstage CLI provides `prepack` and
`postpack` commands that help prepare the package for publishing. These scripts
are automatically run by Yarn before publishing a package.

The `prepack` command will take entry point fields in `"publishConfig"`, such

> as, `"main"` and `"module"`, and move the top level of the `package.json`. This

Reviewer: ```suggestion
as `"main"` and `"module"`, and move them to the top level of the `package.json`. This
```
